### PR TITLE
Refigure gc tests

### PR DESCRIPTION
**Pull request by rotu**
Instead of using `weak-napi`, use the built-in `WeakRef`.
Use the native garbage collector with `{type: 'major', execution: 'async'}`.
If gc is not available and GC tests not explicitly skipped, fail test explicitly.

### CODE REVIEW

File: test/unit/socket-close-test.ts [hunk 1]
```diff
@@ -112,19 +112,13 @@ for (const proto of testProtos("tcp", "ipc", "inproc")) {
         }
         this.slow(200)
 
-        const weak = require("weak-napi") as typeof import("weak-napi")
```
Comment (aminya) [inline test/unit/socket-close-test.ts:115]: Do these tests pass after changing to WeakRef? If so, it might worth it to make a separate PR for this specific change in the tests

Comment (rotu) [reply test/unit/socket-close-test.ts:115]: Lol. I was rebasing to find that out! Lots of the stuff in this pull request has been superceded, but I think there are a few useful things. It's worrying that weak-napi broke the build in the first place though. Any idea why?

Comment (aminya) [reply test/unit/socket-close-test.ts:115]: The issue was caused by outdated dependencies. I forked weak-napi, and now I fixed the tests. The tests seem to be working better than before! So, we might not need these changes
 
https://github.com/zeromq/zeromq.js/pull/623

Comment (aminya) [reply test/unit/socket-close-test.ts:115]: But I'd be still interested in you separating it into another pull request where we can test in isolation.

Comment (rotu) [reply test/unit/socket-close-test.ts:115]: > The issue was caused by outdated dependencies in weak-napi. I forked weak-napi, and now I fixed the tests.

How did the outdated dependencies cause it to fail?

The odd thing is that the `weak-napi` repo and its tests seem to work for me. It only was a problem when included as a dependency of `zeromq.js`.

I'm working on separating it out and I think I figured out the big parts. I like the idea of using standard APIs like `WeakRef` or `FinalizationRegistry` rather than pulling in yet another library!

File: tsconfig.json [hunk 2]
```diff
@@ -1,10 +1,12 @@
 {
   "compilerOptions": {
     "allowJs": false,
-    "target": "es2020",
+    "target": "es2022",
```
Comment (aminya) [inline tsconfig.json:4]: Please revert these ES target changes. Bumping this would mean we will not be able to support some older versions of Nodejs. This should be treated separately in a backwards-compatible way once we move to Es Modules/

**Review by aminya — CHANGES_REQUESTED** [review]

**Comment by rotu** [comment]
> The tests fail with this error:
> 
> ```js
> TypeError: Converting circular structure to JSON
>     --> starting at object with constructor 'SourceFileObject'
>     |     property 'statements' -> object with constructor 'Array'
>     |     index 0 -> object with constructor 'NodeObject'
>     --- property 'parent' closes the circle
>     at stringify (<anonymous>)
>     at writeChannelMessage (node:internal/child_process/serialization:159:20)
>     at process.target._send (node:internal/child_process:845:17)
>     at process.target.send (node:internal/child_process:745:19)
>     at /home/runner/work/zeromq.js/zeromq.js/node_modules/.pnpm/workerpool@6.2.1/node_modules/workerpool/src/worker.js:139:22
> ```

That's interesting.... I can see that here. It's failing when trying to send an error from the Typescript compiler. Curious why this didn't turn up earlier...

```
"⨯ Unable to compile TypeScript:\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m7\u001b[0m:\u001b[93m1\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'describe'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m7\u001b[0m describe(\"context process exit\", function () {\r\n\u001b[7m \u001b[0m \u001b[91m~~~~~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m8\u001b[0m:\u001b[93m3\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'describe'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m8\u001b[0m   describe(\"with default context\", function () {\r\n\u001b[7m \u001b[0m \u001b[91m  ~~~~~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m9\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m9\u001b[0m     it(\"should occur when sockets are closed\", async function () {\r\n\u001b[7m \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m10\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m10\u001b[0m       this.slow(200)\r\n\u001b[7m  \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m21\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m21\u001b[0m     it(\"should occur when sockets are not closed\", async function () {\r\n\u001b[7m  \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m22\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m22\u001b[0m       this.slow(200)\r\n\u001b[7m  \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m31\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m31\u001b[0m     it(\"should not occur when sockets are open and polling\", async function () {\r\n\u001b[7m  \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m32\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m32\u001b[0m       this.slow(1000)\r\n\u001b[7m  \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m45\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m45\u001b[0m     it(\"should produce warning when messages are queued with blocky\", async function () {\r\n\u001b[7m  \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m46\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m46\u001b[0m       this.slow(2000)\r\n\u001b[7m  \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m69\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m69\u001b[0m     it(\"should produce warning when messages are queued without blocky\", async function () {\r\n\u001b[7m  \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m70\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m70\u001b[0m       this.slow(2000)\r\n\u001b[7m  \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m86\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m86\u001b[0m     it(\"should not produce warning when messages are queued for a short time\", async function () {\r\n\u001b[7m  \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m87\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m87\u001b[0m       this.slow(1000)\r\n\u001b[7m  \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m101\u001b[0m:\u001b[93m3\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'describe'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m101\u001b[0m   describe(\"with custom context\", function () {\r\n\u001b[7m   \u001b[0m \u001b[91m  ~~~~~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m102\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m102\u001b[0m     it(\"should occur when sockets are closed\", async function () {\r\n\u001b[7m   \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m103\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m103\u001b[0m       this.slow(200)\r\n\u001b[7m   \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m115\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m115\u001b[0m     it(\"should occur when sockets are closed and context is gced\", async function () {\r\n\u001b[7m   \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m116\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m116\u001b[0m       this.slow(200)\r\n\u001b[7m   \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m133\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m133\u001b[0m     it(\"should occur when sockets are not closed\", async function () {\r\n\u001b[7m   \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m134\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m134\u001b[0m       this.slow(200)\r\n\u001b[7m   \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m144\u001b[0m:\u001b[93m5\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2593: \u001b[0mCannot find name 'it'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha` and then add 'jest' or 'mocha' to the types field in your tsconfig.\r\n\r\n\u001b[7m144\u001b[0m     it(\"should not occur when sockets are open and polling\", async function () {\r\n\u001b[7m   \u001b[0m \u001b[91m    ~~\u001b[0m\r\n\u001b[96mtest/unit/context-process-exit-test.ts\u001b[0m:\u001b[93m145\u001b[0m:\u001b[93m7\u001b[0m - \u001b[91merror\u001b[0m\u001b[90m TS2683: \u001b[0m'this' implicitly has type 'any' because it does not have a type annotation.\r\n\r\n\u001b[7m145\u001b[0m       this.slow(1000)\r\n\u001b[7m   \u001b[0m \u001b[91m      ~~~~\u001b[0m\r\n"
```

**Comment by aminya** [comment]
Yes, please revert any changes other than the weak-napi changes. Let's get the tests running again.

**Comment by aminya** [comment]
I moved these changes to #631
